### ISSUE TITLE
Guaranteed StackOverflowException for subtypes of Map.Entry with self-referential type parameters

### ISSUE THREAD

**Issue by jnape** [FasterXML/jackson-databind#1928]
The following test fails with a stack overflow exception:

```java
    interface Interface<I extends Interface> extends Map.Entry<String, String> {
    }

    class Foo implements Interface<Foo> {
        @Override
        public String getKey() {
            throw new UnsupportedOperationException("not yet implemented");
        }

        @Override
        public String getValue() {
            throw new UnsupportedOperationException("not yet implemented");
        }

        @Override
        public String setValue(String value) {
            throw new UnsupportedOperationException("not yet implemented");
        }
    }

    @Test
    public void shouldAllowSubTypesOfMapEntryWithRecursiveTypeSignatures() {
        JsonSerializer<Foo> fooSerializer = new JsonSerializer<Foo>() {
            @Override
            public void serialize(Foo value, JsonGenerator gen, SerializerProvider serializers) throws IOException, JsonProcessingException {
                gen.writeString("hi");
            }
        };
        ObjectMapper objectMapper = new ObjectMapper().registerModule(new SimpleModule().addSerializer(Foo.class, fooSerializer));
        objectMapper.valueToTree(new Foo());
    }
```

Using a completely vanilla `ObjectMapper` also doesn't help.

Tested with jackson-databind 2.6.4.

**Comment by cowtowncoder** [comment]
@jnape As a general rule, please test bugs with latest patch of the branch (2.6.7 here I think).

But aside that in this particular case testing really is needed with a later version since type introspection was fully rewritten for 2.7. So it'd be good to know it still occurs with, say, latest stable version, 2.9.4.

**Comment by jnape** [comment]
@cowtowncoder Thanks for the quick feedback. Indeed, the issue does not exist in the latest version (2.9.4).

**Comment by cowtowncoder** [comment]
@jnape Ah good! thank you for verifying, and glad this issue amongst others was resolved, probably in 2.7 but possible even later (there were more changes during 2.8.x).
